mindmeld/models/nn_utils/input_encoders.py

```diff
@@ -0,0 +1,937 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This module consists of encoders that serve as input to pytorch modules
+"""
+import json
+import logging
+import os
+from abc import abstractmethod
+from typing import Dict, List, Union, Any, Iterable
+
+from .._util import _get_module_or_attr
+from ..containers import GloVeEmbeddingsContainer, HuggingfaceTransformersContainer
+
+try:
+    import torch
+    from torch.nn.utils.rnn import pad_sequence
+
+    nn_module = _get_module_or_attr("torch.nn", "Module")
+except ImportError:
+    nn_module = object
+    pass
+
+logger = logging.getLogger(__name__)
+
+LABEL_PAD_TOKEN_IDX = -1  # value set based on default label padding idx in pytorch
+
+
+# abstract encoder; derives from torch nn.Module
+
+
+class AbstractEncoder(nn_module):
+
+    def __init__(self):
+        super().__init__()
+        self.name = self.__class__.__name__
+        self.emb_dim = None
+        self.params_keys = set(["name", "emb_dim"])
+
+    def forward(self, *args, **kwargs):
+        return self.batch_encode(*args, **kwargs)
+
+    @abstractmethod
+    def fit(self, **kwargs):
+        msg = f"Subclass {self.name} need to implement this method"
+        raise NotImplementedError(msg)
+
+    @abstractmethod
+    def dump(self, **kwargs):
+        msg = f"Subclass {self.name} need to implement this method"
+        raise NotImplementedError(msg)
+
+    @classmethod
+    @abstractmethod
+    def load(cls, **kwargs):
+        msg = f"Subclass {cls.__name__} need to implement this method"
+        raise NotImplementedError(msg)
+
+    @abstractmethod
+    def batch_encode(self, examples, labels=None, **kwargs) -> Dict[str, Any]:
+        msg = f"Subclass {self.name} need to implement this method"
+        raise NotImplementedError(msg)
+
+    # common `get` methods
+
+    def get_num_tokens(self):
+        if not hasattr(self, "token2idx"):
+            return None
+        return len(self.token2idx)
+
+    def get_emb_dim(self):
+        return self.emb_dim
+
+    def get_pad_token_idx(self):
+        if not hasattr(self, "pad_token_idx"):
+            return None
+        return self.pad_token_idx
+
+    def get_embedding_weights(self):
+        if not hasattr(self, "token2emb") or not self.token2emb:
+            msg = f"Encoder instance ({self.name}) does not contain any token-to-embeddings mapping"
+            logger.info(msg)
+            return None
+        embedding_weights = {}
+        for token, idx in self.token2idx.items():
+            if token in self.token2emb:
```

Comment:
If a word is not present in the embedding dict, shouldn't we initialize a random vector? What happens in case of special tokens/tokens that are user defined?